### PR TITLE
Chat protocol rewrite

### PR DESCRIPTION
**Pull request by nerzhul**
New TOCLIENT_CHAT_MESSAGE packet

* Rename old packet to TOCLIENT_CHAT_MESSAGE_OLD for compat
* Handle TOCLIENT_CHAT_MESSAGE new structure client side

This change will be required by client side modding

* [x] Change chat message queue raw strings to ChatMessageEvent (type, string, sender)

### CODE REVIEW

**Comment by rubenwardy** [comment]
Why can't you just do

```
u16 length
wstring message
u8 version
u8 message_type
u16 sendername length
wstring sendername
```

On the server you write the packet depending on the client, eg for old clients you only send the length and message with the sendername in <>

**Comment by nerzhul** [comment]
@rubenwardy using current packet doesn't permit to use proper wstring encoding/decoding (in NetworkPacket class), look at the decoding part client side. This also permit to fix the chat wstring encoding issue in the current protocol.
Also in protocol terms, having version not at the beginning is just a headhacke. Rework the whole packet is a good thing here, also you will see server side changes to use this packet, but this need a little bit more work than now (we cannot send newer packet)

**Comment by red-001** [comment]
could you also add a timestamp to the chat packet so that something like #3838 can be implemented?

**Comment by ShadowNinja** [comment]
Chat messages should be encoded as UTF-8 and only converted to UTF-16 for display on the GUI (and only then because the Irrlicht API currently requires that).

**Comment by nerzhul** [comment]
i can add a timestamp and remove the wstring if needed (and rebase the PR + remove compat)

**Comment by nerzhul** [comment]
I rebased it and added a timestamp added by server when sending message

**Comment by nerzhul** [comment]
Last remaining point: sending the chat message in its object form to lua instead of sending only raw message or message without other informations

File: src/clientiface.cpp [hunk 1]
```diff
@@ -693,6 +693,28 @@ void ClientInterface::sendToAll(NetworkPacket *pkt)
...
+		if (client->net_proto_version >= min_proto_ver) {
+			m_con->Send(client->peer_id,
+						clientCommandFactoryTable[pkt->getCommand()].channel,
+						pkt,
+						clientCommandFactoryTable[pkt->getCommand()].reliable);
+		} else if (client->net_proto_version != 0) {
+			m_con->Send(client->peer_id,
+						clientCommandFactoryTable[legacypkt->getCommand()].channel,
+						legacypkt,
+						clientCommandFactoryTable[legacypkt->getCommand()].reliable);
```
Comment (SmallJoker) [inline src/clientiface.cpp]: How about:
```C++
RemoteClient *client = i->second;
NetworkPacket *dst_pkt = nullptr;

if (client->net_proto_version >= min_proto_ver)
	dst_pkt = pkt;
else if (client->net_proto_version != 0)
	dst_pkt = legacypkt;
else
	continue;

m_con->Send(client->peer_id,
	clientCommandFactoryTable[dst_pkt->getCommand()].channel,
	dst_pkt,
	clientCommandFactoryTable[dst_pkt->getCommand()].reliable);
```

File: src/network/clientpackethandler.cpp [hunk 2]
```diff
@@ -413,8 +416,44 @@ void Client::handleCommand_ChatMessage(NetworkPacket* pkt)
...
+	*pkt >> chatMessage->sender >> chatMessage->message >> chatMessage->timestamp;
+
+	chatMessage->type = (ChatMessageType) message_type;
+
+	// @TODO send this to CSM using ChatMessage object
+	if (!moddingEnabled() || !m_script->on_receiving_message(
+			wide_to_utf8(chatMessage->message))) {
+		pushToChatQueue(chatMessage);
+	}
+	else {
```
Comment (SmallJoker) [inline src/network/clientpackethandler.cpp]: `} else {` or no brackets at all

File: src/network/networkprotocol.h [hunk 3]
```diff
@@ -307,7 +315,17 @@ enum ToClientCommand
...
+		u8 version
+		u8 message_type
+		u16 sendername length
+		wstring sendername
+		u16 length
+		wstring message
+	*/
+
+	TOCLIENT_CHAT_MESSAGE_OLD = 0x30,
 	/*
```
Comment (SmallJoker) [inline src/network/networkprotocol.h]: A comment here would be nice: `Deprecated since proto ver 35`

File: src/server.cpp [hunk 4]
```diff
@@ -1615,21 +1617,30 @@ void Server::SendInventory(PlayerSAO* playerSAO)
...
 		else
-			pkt << message;
-
-		Send(&pkt);
-	} else {
-		for (u16 id : m_clients.getClientIDs())
-			SendChatMessage(id, message);
+			Send(&pkt);
+	}
+	else {
```
Comment (SmallJoker) [inline src/server.cpp]: `} else {`

File: src/network/networkprotocol.h (hunk 3 above)
Comment (nerzhul) [inline src/network/networkprotocol.h]: sure

File: src/client.h [hunk 5]
```diff
@@ -328,6 +329,7 @@ class Client : public con::PeerHandler, public InventoryManager, public IGameDef
 	void handleCommand_BlockData(NetworkPacket* pkt);
 	void handleCommand_Inventory(NetworkPacket* pkt);
 	void handleCommand_TimeOfDay(NetworkPacket* pkt);
+	void handleCommand_ChatMessageOld(NetworkPacket* pkt);
```
Comment (SmallJoker) [inline src/client.h]: `NetworkPacket *pkt`

File: src/clientiface.cpp [hunk 6]
```diff
@@ -693,6 +693,35 @@ void ClientInterface::sendToAll(NetworkPacket *pkt)
...
+		if (client->net_proto_version >= min_proto_ver) {
+			pkt_to_send = pkt;
+		} else if (client->net_proto_version != 0) {
+			pkt_to_send = legacypkt;
+		}
+		else {
+			std::stringstream ss;
+			ss << "Client with unhandled version to handle: '"
+				<< client->net_proto_version << "'";
+			FATAL_ERROR(ss.str().c_str());
```
Comment (SmallJoker) [inline src/clientiface.cpp]: An invalid protocol version shouldn't trigger a "crash".

File: src/clientiface.cpp [hunk 7]
```diff
@@ -693,6 +693,35 @@ void ClientInterface::sendToAll(NetworkPacket *pkt)
...
+			i != m_clients.end(); ++i) {
+		RemoteClient *client = i->second;
+		NetworkPacket *pkt_to_send = nullptr;
+
+		if (client->net_proto_version >= min_proto_ver) {
+			pkt_to_send = pkt;
+		} else if (client->net_proto_version != 0) {
+			pkt_to_send = legacypkt;
+		}
+		else {
```
Comment (SmallJoker) [inline src/clientiface.cpp]: `} else {` (perhaps wrong auto-formatting?)

File: src/clientiface.cpp (hunk 6 above)
Comment (nerzhul) [inline src/clientiface.cpp]: exact

File: src/network/networkprotocol.h [hunk 8]
```diff
@@ -307,7 +315,17 @@ enum ToClientCommand
...
+	/*
+		u8 version
+		u8 message_type
+		u16 sendername length
+		wstring sendername
+		u16 length
+		wstring message
+	*/
+
+	TOCLIENT_CHAT_MESSAGE_OLD = 0x30, // Deprecated by proto v25
```
Comment (SmallJoker) [inline src/network/networkprotocol.h]: *v35

Comment (nerzhul) [inline src/network/networkprotocol.h]: exact and maybe v35, depending on the flavours

**Comment by SmallJoker** [comment]
Compatibility tested:
- [x] Old server & new client
- [x] New server & old client
- [x] New server & new client

Works. :+1:

**Pull request merged**